substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/sources/Source.java

```diff
@@ -0,0 +1,143 @@
+     * <p>
+     * As an optimization, the lookup thread may transition to an intermediate state,
+     * <code>LOCATED</code> after a source is located but before it is copied into the source cache.
+     * It will notify all waiting threads after this transition, allowing them to return early.
+     */
+    private CacheStatus status;
+
+    Source(Path path) {
+        this.path = path;
+        this.status = CacheStatus.UNKNOWN;
+    }
+
+    /*
+     * n.b. the following accessors are deliberately not synchronized. In some cases we don't care
+     * about racy reads. In others we do. Use with due care.
+     */
+    public boolean isUnknown() {
+        return status == CacheStatus.UNKNOWN;
+    }
+
+    public boolean isCaching() {
+        return status == CacheStatus.CACHING;
+    }
+
+    public boolean isMissing() {
+        return status == CacheStatus.MISSING;
+    }
+
+    public boolean isLocated() {
+        return status == CacheStatus.LOCATED;
+    }
+
+    public boolean isCached() {
+        return status == CacheStatus.CACHED;
+    }
+
+    public boolean isResolved() {
+        return isMissing() || isLocated() || isCached();
+    }
+
+    public Path getPath() {
+        return (isMissing() ? null : path);
+    }
+
+    public synchronized void updateStatus(CacheStatus status) {
+        /* state changes must be synchronized and must wake up *all* waiting threads */
+        this.status = status;
+        notifyAll();
+    }
+
+    /**
+     * Check the status of a source file to determine whether the caller should try to cache it or
+     * not. For any given source only one caller should be requested to attempt to cache the file.
+     * The method is synchronized in order to avoid races to cache the source. If a cache attempt is
+     * in progress at the point of call the caller will wait for that attempt to be resolved before
+     * returning. A call is allowed to return at the point where the source is located but before it
+     * is actually copied into the cache.
+     *
+     * The status transitions are relatively simple. A new source i.e. one with status UNKNOWN is
+     * transitioned to CACHING before returning true. In this case the caller should attempt to
+     * locate and cache the source file. That is the only transition performed by this method.
+     *
+     * The caller that attempts the cache operation performs all other transitions. It upgrades the
+     * status to LOCATED when it finds a matching source and then to CACHED when this file is copied
+     * into the cache. It can also upgrade the status from CACHING to MISSING if the file cannot be
+     * found or downgrade it from LOCATED to MISSING if an attempt to copy a located source into the
+     * cache fails.
+     *
+     * @return True if the calling thread is expected to try to cache the source otherwise false.
+     */
+    synchronized boolean shouldCache() {
```

Comment:
The name `shouldCache` implies that this method does far less than what it actually does. Unfortunately, I can't think of any better one. Maybe `shouldCacheOrBlock`.